### ISSUE TITLE
Delay (Text with effects)?

### ISSUE THREAD

**Issue by Cerzs** [bbepis/XUnity.AutoTranslator#35]
Hi there, I noticed that the text is delayed when there are effects, although _AutoGeneratedTranslations.{lang}.txt exists. Any way to avoid it? (For redistributing)

![lHLnCZN](https://user-images.githubusercontent.com/44489935/59384392-0d253680-8d30-11e9-849b-1fa2487f778c.gif)

**Comment by gravydevsupreme** [comment]
This is certainly expected behavior at the moment. 

Currently, the only way to (potentially) solve it is to provide partial translations for the sentence "scrolling" in. However, even that would not be guaranteed to work, as it would then be dependent on how the game engine updates the text in the text component.

If you tell me what game it is, then I can try to test out some things.

**Comment by Cerzs** [comment]
Thanks for answering, @gravydevsupreme. The game is [Broke Girl](https://www.dlsite.com/ecchi-eng/work/=/product_id/RE249278.html). 

It occurs with any game with that effect. Can be controlled just changing in-game configuration (text speed), but some games does not have config menu.

**Comment by gravydevsupreme** [comment]
You can try out this version and report back if it works for you.

You have to set configuration, like so:

```ini
[Behaviour]
GeneratePartialTranslations=False
```

(irrelevant configuration left out)

[dist-custom.zip](https://github.com/bbepis/XUnity.AutoTranslator/files/3311116/dist-custom.zip)

It may increase the memory usage of the plugin if there is a large amount of translations, though.

**Comment by Cerzs** [comment]
Hi! It works, but I have a problem. This version takes the "you" (name) and "you" from "you know" and replaces the name. 

"You" know = Sabes (Sab) 

![XD4p2U8](https://user-images.githubusercontent.com/44489935/59899438-b28c7a00-93c2-11e9-86f2-1024addd65db.jpg) 
![XB2MTkx](https://user-images.githubusercontent.com/44489935/59900332-e87f2d80-93c5-11e9-9579-406cfcfb2e65.jpg)


Also, sometimes, it cuts the sentences because there's one similar. Example:

Who are you? Why are you here? = ¿Quién eres tú? ¿Qué haces aquí?
Then:
Who are you? = ¿Quién eres t (This is not in the .txt, it took the previous "who are you?" and applies the "**quién eres t**")

![PvY4ba5](https://user-images.githubusercontent.com/44489935/59900152-39425680-93c5-11e9-9d09-e66f8df67343.jpg)

**Comment by gravydevsupreme** [comment]
I'm not entirely sure what you mean by the first issue, could you rephrase that? Also, is this only an issue in this version? It doesn't sound like the changes I made in this version would affect that.

The second issue you mention can only be avoided if you do not enable the feature while actually translating the game.

If you enable after you have it translated, this type of issue should not appear as then the full translations will take priority over partial ones.

This, of course, could mean that text sometimes feel a bit "janky" as it scrolls in, if there is a full translation that takes priority over a partial one.

But there is really no way to avoid that without directly taking a dependency on the specific game.

**Comment by gravydevsupreme** [comment]
Woopsy. After looking a bit further into the custom distribution I sent, I can see that the way I implemented actually caused partial translations to override full translations, which were not intentional and may have caused the first issue you mentioned.

What I mentioned for the second one still stands, though. Here's another version you can try.

[custom-dist2.zip](https://github.com/bbepis/XUnity.AutoTranslator/files/3315810/custom-dist2.zip)

**Comment by Cerzs** [comment]
This works perfect, both issues are fixed. Thanks! 

Btw, last question, is not about this issue. Would be possible to hook text from another location? Because I have a game (Unity, of course) that has the text files inside a .pac. 

![akQHmmK](https://user-images.githubusercontent.com/44489935/59950381-c1674100-9443-11e9-9016-edfb84ce7c55.jpg)

**Comment by gravydevsupreme** [comment]
If you have an inquiry about a specific game, I do not mind taking a look at why it doesn't work.

I don't know what a .pac file is, though. Nor would it be what this plugin hooks. This plugin does not hook resource files (if that is what it is), but rather managed code in order to intercept text as it is displayed.

**Comment by Cerzs** [comment]
I see. If you can, take a look, please. The game is [Kara no Shoujo The Second Episode](https://sukebei.nyaa.si/view/2509212). The .pac contains .txt files.

Thanks again.

**Comment by gravydevsupreme** [comment]
Looked into the game. It uses a ADV-engine that I have not seen before and a positively ancient version of TextMeshPro.

I have also gotten it to work, but there are some gotcha to consider:
 *  due to its fairly heavy use of 'rich text' (text with markup), I've had to change how the plugin handles rich text in general (for the better). But there's one issue you need to be aware of if you are doing manual translations (in this game):
    * The text displayed in the backlog is not the same as the text in the 'story display' as the game adds a '<#000000>' in front of all of it to change its color. Translations with these tags in front are not needed because the plugin is smart enough to consider the text without the tag. But they will be generated and added to the translation file IF you enter the backlog and the text from the 'story display' has not already been translated. In general, you will be able to just remove these tags in the translation file if they are generated.
 * The font used in the game supports a very limited character set, probably even a subset of ASCII. Any character that is not supported will be shown as a big red 'X'. So you either have to avoid these special characters or change the font in the game through the option, `OverrideFontTextMeshPro=Fonts & Materials/ARIAL SDF`. For TextMeshPro fonts have to be build specifically for the game, so I do not think there is an easy to to add an alternative font other than the standard 'Arial' font that comes with it.
 * The game uses special keywords that enables lookups in a glossary. These keywords must appear in the translated text exactly as they are in original (I think). Also, if the translation happens asynchronously (not an issue for manual translations), then the keyword highlight may be offset in relation to the actual text. It is probably possible to fix this, but for now, my attempts were in vain.
 * Other issues? Perhaps. You would have to tell me. :)
 
With that being said, if what you are looking for is to simply provide a full manual translation of the game, then it may also be possible to just edit those ".pac" files you mentioned (although I do not know what tooling you would use to unpack/pack them, but from what I read in the source code of the ADV-engine, it looks fairly simple (encrypted with a simple XOR-cipher with the key hardcoded in the source code)).

Here's a version you can test against this game.

[dist-custom3.zip](https://github.com/bbepis/XUnity.AutoTranslator/files/3317930/dist.zip)

**Comment by Cerzs** [comment]
The Arial font works perfect with the IPA plugin. I prefer BepInEx, tho. Because that cmd that appears before launching. Thanks, really. A friend wanted to translate this game for almost 3 years.

Also, yes. I can extract/repack the .pac files, but the we had the problem with the font. If you need the unpacker/repacker for something, let me know.

If you figure out why with BepInEx crashes and found a fix, let me know please. 

Thanks again!

**Comment by denikson** [comment]
@gravydevsupreme 
Some older Unity versions require you to adjust the entrypoint of BepInEx plugin loading.  
In the case of Kara no Shoujo The Second Episode, it seems that you need to adjust the entrypoint as follows in `BepInEx.cfg`:

```ini
[Preloader.Entrypoint]

# The local filename of the assembly to target.
Assembly = UnityEngine.dll

# The name of the type in the entrypoint assembly to search for the entrypoint method.
Type = MonoBehaviour

# The name of the method in the specified entrypoint assembly and type to hook and load Chainloader from.
Method = .cctor
```

**Comment by Cerzs** [comment]
Btw, if I can edit the text inside the scripts.pac, then is not neccesary the plugin for translation, right? Just for redirect the font to Arial.

**Comment by gravydevsupreme** [comment]
Can confirm BepInEx 5.0 works with @denikson's config changes.

@Cerzs I can imagine that you could probably do it entirely through changing of the ".pac" files, but I do not know, as I have no idea about their structure and how they all bind the "scenario" together.

I can imagine both that it could be done, but also that it would be easy to break the game entirely by making a small mistake.

If you simply use the plugin for updating the font, you probably also need to unconfigure the endpoint `Endpoint=`, as otherwise the plugin will assume any text it finds is in the "original" language because it has no proper way of knowing whether or not a text/string is translated or not as both languages uses the same character set.

**Comment by Cerzs** [comment]
Yep, works with denikson's config. Thanks! I'll test with both methods and choose the less broken.

Just in case, this game has a prequel called Cartagra, I checked it, only has the same limited font (rosarivo) but not Arial or similars. I can presume that work with it is almost impossible without figure out how to import another font, right?

**Comment by gravydevsupreme** [comment]
Yes. It seems that the method used to load the image into a texture does not exist in this version (5.2) of Unity.

```
[Error  :XUnity.AutoTranslator 3.6.1] An error occurred while translating texture.
System.MissingMethodException: Method not found: 'UnityEngine.Texture2D.LoadImage'.
  at XUnity.AutoTranslator.Plugin.Core.Extensions.TextureExtensions.LoadImageEx (UnityEngine.Texture2D texture, System.Byte[] data, Boolean markNonReadable) [0x00000] in <filename unknown>:0
  at XUnity.AutoTranslator.Plugin.Core.AutoTranslationPlugin.TranslateTexture (System.Object source, UnityEngine.Texture2D texture, Boolean isPrefixHooked, XUnity.AutoTranslator.Plugin.Core.TextureReloadContext context) [0x00000] in <filename unknown>:0
  at XUnity.AutoTranslator.Plugin.Core.AutoTranslationPlugin.HandleImage (System.Object source, UnityEngine.Texture2D texture, Boolean isPrefixHooked) [0x00000] in <filename unknown>:0
```

I'll take a look to see what alternatives were offered back then to see if it is possible to support it.

As for the other game, I can take a look. 

I know the tooling provided in the editor for TextMeshPro enables you to create fonts for your games, but I've never really used it (as I have never actually used the Unity Editor for anything serious). I can imagine it is probably possible to use this tooling at runtime to generate a custom font based on a "standard font".

Usually these default resources are exported together with the game, though. So I can only imagine there is some alternative embedded in the game files.

**Comment by gravydevsupreme** [comment]
First things first:

Texture's in Kara no Shoujo 2nd:
 * I got a version that "somewhat" fixes the texture issue. Even though I thought it would be completely straightforward it was not because the game has a tendency to screw up image loading and I have absolutely no idea why. I've introduced an option `EnableLegacyTextureLoading=True` (defaults to False), which fixes the issue in most cases, but I have no idea why it fixes it in most cases. I essentially just load the texture twice and then it works... Whaaaa... Anyway, if you have translated an image and it shows up as white with red stripe(s), you should just run away screaming, or perhaps ask yourself if translating that image is really that important. 
 * Fixed a bug with partial translation introduced in last distribution I shared
 
As for 'Cartagra', this seems to use an even more ancient version of Unity (4.5). I've fixed the most noticable things related to this engine, so now it is actually capable capturing and translating shown texts, but that also really it for the positive news:
 * This game specifically seems to want to translate texts that breaks the way I encode/decode texts into the translations file. The way I am doing this at the moment is arguably pretty trash, so I am considering a way of fixing it, but this is not fixed yet, so you may find the plugin sometimes re-translating a seemingly translated text.
 * The game does not seem to have a built-in font, like the other game, even though a resource is referenced as "Fonts/Arial SDF" in the TMP decompiled source code. I am not currently going to investigate this issue further. To put it shortly, too much effort. :)
 * It also uses a different version of the before-mentioned ADV engine, which causes issues with some text being cut off. I can probably fix this if important.
 * It also shines a bright line on a whitespacing issue that exists in the plugin, which is already on my TODO list to fix. This is because the ADV-engine uses the same textbox for names and dialogue.
 * (There will also be an error in the console during startup, but this can be safely ignored as it is only related to the translation aggregator view)

In the plugin's current state, I would not recommend trying to create a manual translation for 'Cartagra' with it.

[dist-custom4.zip](https://github.com/bbepis/XUnity.AutoTranslator/files/3322377/dist-custom4.zip)

**Comment by Cerzs** [comment]
Thanks for fix KnS. It works! I'll consider about running away if the problem you mencioned appears, haha.

And wow, Cartagra sounds like a pain, lol. I really apreciate your efforts, don't do anything if is too much problems for both sides. Thanks again.

**Comment by gravydevsupreme** [comment]
The latest version of XUA (4.7.0) now has a function that allows preprocessing of text before sending it to the translator. It is a special file (like substitutions) that you can put key-value pairs into. By default the file is called `_Preprocessors.txt`.

You can put in the entry:

```
\\n=\n
```

And the game will then properly handle newlines, as far as I can see.
